pkg/pattern/drain/line_tokenizer.go

```diff
@@ -131,3 +174,115 @@ func (splittingTokenizer) Join(tokens []string, state interface{}) string {
 	}
 	return strBuilder.String()
 }
+
+func (splittingTokenizer) Clone(tokens []string, state interface{}) ([]string, interface{}) {
+	res := make([]string, len(tokens))
+	for i, token := range tokens {
+		res[i] = strings.Clone(token)
+	}
+	if state == nil {
+		return res, nil
+	}
+	spacesAfter := state.([]int)
+	spacesAfterCopy := make([]int, len(spacesAfter))
+	copy(spacesAfterCopy, spacesAfter)
+	return res, spacesAfterCopy
+}
+
+type logfmtTokenizer struct {
+	dec        *logfmt.Decoder
+	varReplace string
+}
+
+func newLogfmtTokenizer(varReplace string) *logfmtTokenizer {
+	return &logfmtTokenizer{
+		dec:        logfmt.NewDecoder(nil),
+		varReplace: varReplace,
+	}
+}
+
+func (t *logfmtTokenizer) Tokenize(line string, tokens []string, _ interface{}) ([]string, interface{}) {
+	if cap(tokens) == 0 {
+		tokens = make([]string, 0, 64)
+	}
+	tokens = tokens[:0]
+	t.dec.Reset(unsafeBytes(line))
+	for !t.dec.EOL() && t.dec.ScanKeyval() {
+		key := t.dec.Key()
+		if isVariableField(key) {
+			tokens = append(tokens, unsafeString(t.dec.Key()), t.varReplace)
+
+			continue
+		}
+		// todo we want to pass bytes and let user copy if needed.
+		tokens = append(tokens, unsafeString(t.dec.Key()), unsafeString(t.dec.Value()))
+	}
+	if t.dec.Err() != nil {
+		return nil, nil
+	}
+	return tokens, nil
+}
+
+func (t *logfmtTokenizer) Join(tokens []string, _ interface{}) string {
+	if len(tokens) == 0 {
+		return ""
+	}
+	if len(tokens)%2 == 1 {
+		tokens = append(tokens, "")
+	}
+	buf := bytes.NewBuffer(make([]byte, 0, 1024))
+	enc := gologfmt.NewEncoder(buf)
+	for i := 0; i < len(tokens); i += 2 {
+		k, v := tokens[i], tokens[i+1]
```

Comment:
How does this handle multi-word values? Are they a single token?

Reply:
correct they are.